lua filter: ensure active_request_ is verified before deletion Fixes #38017

leocfig opened the pull request:
Commit Message:
Lua filter was removing ":status" and then setting a response body, which was causing envoy to terminate with a SIGSEGV. The issue occurred because "onEncodeComplete()" attempted to delete "activerequest" via "deferredDelete()", but "activerequest" had already been deleted earlier, leading to a segmentation fault. This fix ensures that "activerequest" is only deleted once, preventing double deletion. This is done by verifying that "activerequest" is not null before the existing check of "activerequest->remotecomplete".

Additional Description:
I compiled Envoy in release mode (using the -c flag) and also ran the new test in release mode to verify the fix.

Risk Level: Low

Testing:
I added a new integration test in "lua_integration_test.cc" file specifically for this bug. Without the fix, the test fails; with it applied, the test passes successfully.

Fixes #38017

leocfig commented:
I believe the pipeline is failing because the tests are running in debug mode. In this mode, the code eventually reaches a verification where [ENVOY_BUG()](https://github.com/envoyproxy/envoy/blob/6c38b6a020e7577ccffff9e56c7b1b88cd9056a9/source/common/http/utility.cc#L613) forcefully aborts the program, when in **debug** mode. What my fix avoids is the program ending with a segmentation fault in **release** mode.

KBaichoo commented:
Hey @leocfig ,

Thank you for working on this.

We do run CI (strangely under pre-check) with the release mode run of tests: https://github.com/envoyproxy/envoy/actions/runs/14021237617/job/39253502671

It seems like the test is still failing in release mode. 

You can also try wrapping the test with a macro similar to https://github.com/envoyproxy/envoy/blob/2533c82d653c3da2c3f69568ab8478ce55a1051b/test/test_common/utility.h#L99 to catch the error depending on DEBUG vs release build

/wait

leocfig commented:
Thanks for the suggestion! I’ll first investigate why it’s still failing in release mode since it passed locally, then update the test to use EXPECT_ENVOY_BUG.

leocfig commented:
I realized that my test was failing in the pipeline despite passing locally because my code was not up-to-date with a recent commit that modified the `lua_integration_test.cc` file. After reviewing the changes to this file, I updated my test accordingly.

Additionally, I added the `#ifdef NDEBUG` preprocessor directive to ensure the test only runs in release mode. This was necessary because in **debug** mode, the program aborts before reaching the segmentation fault, making the test irrelevant in that mode.

leocfig commented:
I noticed that the `//test/extensions/filters/http/dynamic_forward_proxy:legacy_proxy_filter_integration_test` target is failing in the pipeline, even though it passes when I run it locally. However, I believe this issue is unrelated to my null verification.

KBaichoo commented:
/wait-any

There's some conversation on slack on how to possibly better address this as it slightly circumvents the fix from https://github.com/envoyproxy/envoy/pull/16745

leocfig commented:
I’ve implemented the suggested changes, and it works as expected! I’ve also left a message on Slack with further details.

KBaichoo commented:
hey @leocfig ,

did you push the changes you referenced in https://github.com/envoyproxy/envoy/pull/38863#issuecomment-2787511347?

leocfig commented:
Hey! Yes, I’ve just now pushed the changes mentioned in [#38863(comment)](https://github.com/envoyproxy/envoy/pull/38863#issuecomment-2787511347). Let me know if anything looks off or needs adjustment.

KBaichoo commented on source/common/http/filter_manager.cc:
```diff
@@ -1224,6 +1225,27 @@ void FilterManager::maybeContinueEncoding(StreamEncoderFilters::Iterator continu
+    const auto status = HeaderUtility::checkRequiredResponseHeaders(headers);
+    if (!status.ok()) {
+      // If the check failed, then we reply with BadGateway, and stop the further processing.
+      sendLocalReply(
```
this might be a problem if the filter already sent a local reply e.g. prior we only did this if we made it here https://github.com/envoyproxy/envoy/pull/38863/files#diff-e91327a311d8130daac76a7c43dca2ae70d5a6a37dc7fdb658cf25b2fde88cc9R1309 but we might now exit here: https://github.com/envoyproxy/envoy/pull/38863/files#diff-e91327a311d8130daac76a7c43dca2ae70d5a6a37dc7fdb658cf25b2fde88cc9R1300

so we might want to modify the filter_manager a bit more to ensure we don't end up sending multiple of these.

leocfig replied:
Good point, I’ll take a closer look at this and follow up with a fix. Thank you for the help!

leocfig replied:
I've been trying to fix the issue, but I'm not sure about this part — do we want to remove the `sendLocalReply` call that happens inside the `if (!continue_iteration)` block? Or should we keep it there and just make sure we don't end up calling `sendLocalReply` again later on, in case it does enter the `if (!continue_iteration)` block?

KBaichoo replied:
I think what we want is to only call sendLocalReply here if sendLocalReply has not already been called https://github.com/envoyproxy/envoy/blob/3be167b56628d6eb1564c2db2826b3a485e4cdaf/source/common/http/filter_manager.cc#L961

leocfig replied:
Perhaps we could add a bool to check for that case. This bool would be added in [here](https://github.com/envoyproxy/envoy/blob/3be167b56628d6eb1564c2db2826b3a485e4cdaf/source/common/http/filter_manager.h#L963), like for example `bool local_reply_sent_{}` and we would only call `sendLocalReply` in the cleanup function if `local_reply_sent_` is false.

KBaichoo reviewed (commented):
/wait

KBaichoo commented on source/common/http/filter_manager.cc:
```diff
@@ -1236,12 +1236,13 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
           Http::Code::BadGateway, status.message(), nullptr, absl::nullopt,
           absl::StrCat(StreamInfo::ResponseCodeDetails::get().FilterRemovedRequiredResponseHeaders,
                        "{", StringUtil::replaceAllEmptySpace(status.message()), "}"));
+      state_.local_reply_sent_ = true;
```
I think we want to set this here https://github.com/envoyproxy/envoy/blob/3be167b56628d6eb1564c2db2826b3a485e4cdaf/source/common/http/filter_manager.cc#L968 -- otherwise if a filter sends the local reply we'd miss it.

leocfig replied:
Thanks for pointing that out! I've updated the code accordingly.

KBaichoo reviewed (commented):
Sorry for the delay -- I think this is nearly there. This is a very complex part of the code, so nice job tackling some of the intricacies here.

/wait

KBaichoo commented:
several tests seem to be meaningfully failing: https://github.com/envoyproxy/envoy/actions/runs/14775459103/job/41482826530

/wait

leocfig commented:
Yes, it looks like the latest change introduced some compile-time errors. Would you recommend reverting it for now, or should I try to fix it directly?

KBaichoo commented:
Yes, it looks like the latest change introduced some compile-time errors. Would you recommend reverting it for now, or should I try to fix it directly?
> It seems like these are test failures failing during running across a slew of builds. I think if you reverted the changes to signal state to avoid multiple local replies we'd _likely_ have a bug there.

leocfig commented:
Thanks for the clarification. Just to confirm, would you recommend exploring an alternative approach to adding the `local_reply_sent_` boolean, or is that direction generally fine and I should focus on fixing the current test failures?

KBaichoo commented:
>Just to confirm, would you recommend exploring an alternative approach to adding the local_reply_sent_ boolean, or is that direction generally fine and I should focus on fixing the current test failures?

If you can think of a better fix that covers the cases we've discussed without the tradeoffs that sgtm. It's a bit hard to know without digging into the tests more if we've violating and assumption of the test or if we've introduced a regression.

leocfig commented:
Understood, I’ll take a closer look and explore if there’s a cleaner alternative.

leocfig commented:
I moved the assignment of` local_reply_sent_ = true` outside of `sendLocalReply()`, ensuring it's set earlier in the control flow. This is necessary because the code path may invoke `sendLocalReply()` before encodeHeaders() is reached, leading to inconsistent state handling otherwise and regression.

Additionally, I addressed the segmentation fault in the test by safely accessing `responseHeaders()` using `.ref`() only after checking `has_value()`, inside the `MakeCleanup` callback.

KBaichoo commented:
nice work, I'll circle back here tomorrow to review this. Test failure doesn't look related.

/retest

leocfig commented:
I've incorporated the requested changes, please let me know if they're in line with what you had in mind. If anything still needs adjusting, I'm happy to update it. Thanks again for your time reviewing this!

KBaichoo reviewed (approved):
Thanks this LGTM. We might want to wait until after the next release to merge just incase something was missed.

/assign @yanavlasov 

for senior maintainer signoff

KBaichoo commented:
/retest
